### PR TITLE
Build script detection takes scripting language providers into account

### PR DESCRIPTION
**Pull request by eskatos**
This PR allows Gradle to detect build scripts according to available scripting languages.

It builds upon the `ScriptingLanguage` SPI introduced by #1908.

See https://github.com/gradle/gradle-script-kotlin/issues/37

### Gradle Core Team Checklist
- [ ] Verify test coverage and CI build status https://builds.gradle.org/viewQueued.html?itemId=2858373
- [x] Verify documentation including proper use of `@since` and `@Incubating` annotations for all public APIs

### CODE REVIEW

**Comment by eskatos** [comment]
@bamboo, updated according to our chat, could you have another look?

File: subprojects/core/src/main/java/org/gradle/testfixtures/internal/ProjectBuilderImpl.java [hunk 1]
```diff
@@ -83,7 +84,7 @@ public Project createProject(String name, File inputProjectDir, File gradleUserH
         GradleInternal gradle = CLASS_GENERATOR.newInstance(DefaultGradle.class, null, startParameter, topLevelRegistry.get(ServiceRegistryFactory.class));
 
         DefaultProjectDescriptor projectDescriptor = new DefaultProjectDescriptor(null, name, projectDir, new DefaultProjectDescriptorRegistry(),
-                topLevelRegistry.get(FileResolver.class));
+                topLevelRegistry.get(FileResolver.class), DefaultScriptFileResolver.empty());
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/testfixtures/internal/ProjectBuilderImpl.java]: :x: I would avoid spreading the knowledge about `DefaultScriptFileResolver` around by making the argument optional as in f8716bb3424b76409c32ab4c3283f2d459a1e0be.

File: subprojects/core/src/main/java/org/gradle/initialization/DefaultProjectDescriptor.java [hunk 2]
```diff
@@ -105,6 +112,16 @@ void setPath(Path path) {
     }
 
     public String getBuildFileName() {
+        if (buildFileName == null) {
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/initialization/DefaultProjectDescriptor.java]: :x: I would change the code to extract the file name from the build file instead. See 0d0e83d837ef877e8205cb9200ceda4aa2e66962 for the rationale.

**Comment by eskatos** [comment]
@bamboo, updated, could you have another look?

**Comment by eskatos** [comment]
CI approved the changes ✅ https://builds.gradle.org/viewLog.html?buildId=2859496
Merging

**Comment by bamboo** [comment]
\o/

**Pull request merged**